src/app/feeds.py

```diff
@@ -0,0 +1,163 @@
+    for entry in feed_data.entries:
+        if entry.id not in existing_posts:
+            logger.debug(f"found new podcast: {entry.title}")
+            p = make_post(feed, entry)
+            # do not allow automatic download of any backcatalog added to the feed
+            if (
+                oldest_post is not None
+                and p.release_date.date() < oldest_post.release_date
+            ):
+                p.whitelisted = False
+                logger.debug(
+                    f"skipping post from archive due to \
+number_of_episodes_to_whitelist_from_archive_of_new_feed setting: {entry.title}"
+                )
+            else:
+                p.whitelisted = config.automatically_whitelist_new_episodes
+            db.session.add(p)
+    db.session.commit()
+    logger.info(f"Feed with ID: {feed.id} refreshed")
+
+
+def add_or_refresh_feed(url: str) -> Feed:
+    feed_data = fetch_feed(url)
+    if "title" not in feed_data.feed:
+        logger.error("Invalid feed URL")
+        raise ValueError(f"Invalid feed URL: {url}")
+
+    feed = Feed.query.filter_by(rss_url=url).first()
+    if feed:
+        refresh_feed(feed)
+    else:
+        feed = add_feed(feed_data)
+    return feed  # type: ignore[no-any-return]
+
+
+def add_feed(feed_data: feedparser.FeedParserDict) -> Feed:
+    logger.info(f"Storing feed: {feed_data.feed.title}")
+    try:
+        feed = Feed(
+            title=feed_data.feed.title,
+            description=feed_data.feed.get("description", ""),
+            author=feed_data.feed.get("author", ""),
+            rss_url=feed_data.href,
+        )
+        db.session.add(feed)
+        db.session.commit()
+
+        num_posts_added = 0
+        for entry in feed_data.entries:
+            p = make_post(feed, entry)
+            if (
+                config.number_of_episodes_to_whitelist_from_archive_of_new_feed
+                is not None
+                and num_posts_added
+                >= config.number_of_episodes_to_whitelist_from_archive_of_new_feed
+            ):
+                logger.info(
+                    f"Number of episodes to load from archive reached: {num_posts_added}"
+                )
+                p.whitelisted = False
+            else:
+                num_posts_added += 1
+                p.whitelisted = config.automatically_whitelist_new_episodes
+            db.session.add(p)
+        db.session.commit()
+        logger.info(f"Feed stored with ID: {feed.id}")
+        return feed
+    except Exception as e:
+        logger.error(f"Failed to store feed: {e}")
+        db.session.rollback()
+        raise e
+
+
+def generate_feed_xml(feed: Feed) -> Any:
+    logger.info(f"Generating XML for feed with ID: {feed.id}")
+    items = []
+    for post in feed.posts:  # type: ignore[attr-defined]
+        items.append(
+            PyRSS2Gen.RSSItem(
+                title=post.title,
+                link=url_for("main.download_post", p_guid=post.guid, _external=True),
```

Comment:
I realized there is a bug when using the server through my reverse proxy, I think due to this line.

Compare:
https://github.com/jdrbc/podly_pure_podcasts/blob/main/src/app/old_routes.py#L124

Reply:
fixed!